tests/unit/test_extractor.py

```diff
@@ -0,0 +1,125 @@
+
+
+def test_write_time_series_to_cdf(extractor, mocker):
+    data = {
+        "externalId": ["id1", "id2", "id1", "id2"],
+        "timestamp": [
+            "2022-02-07T16:01:27.001Z",
+            "2022-02-07T16:03:47.000Z",
+            "2022-02-07T16:03:52.000Z",
+            "2022-02-07T16:04:12.000Z",
+        ],
+        "value": [1.0, 2.0, 3.0, 4.0],
+    }
+    df = pd.DataFrame(data)
+
+    mocker.patch.object(extractor.state_store, 'get_state', return_value=(None, ))
+    mocker.patch.object(extractor.client.time_series.data, 'insert_dataframe', return_value=None)
+    mocker.patch("cdf_fabric_replicator.extractor.CdfFabricExtractor.set_state", return_value=None)
+
+    extractor.write_time_series_to_cdf(df)
+
+    extractor.state_store.get_state.assert_any_call('/table/path-id1-state')
+    extractor.state_store.get_state.assert_any_call('/table/path-id2-state')
+    assert(extractor.client.time_series.data.insert_dataframe.call_count == 2)
+    extractor.set_state.assert_any_call("/table/path-id1-state", df[df["externalId"] == "id1"]["timestamp"].max())
+    extractor.set_state.assert_any_call("/table/path-id2-state", df[df["externalId"] == "id2"]["timestamp"].max())
+
+
+def test_write_time_series_to_cdf_filter_old_data_points(extractor, mocker):
+    data = {
+        "externalId": ["id1", "id1", "id1", "id1"],
+        "timestamp": [
+            "2022-02-07T16:01:27.001Z",
+            "2022-02-07T16:03:47.000Z",
+            "2022-02-07T16:03:52.000Z",
+            "2022-02-07T16:04:12.000Z",
+        ],
+        "value": [1.0, 2.0, 3.0, 4.0],
+    }
+    df = pd.DataFrame(data)
+
+    last_update_time = "2022-02-07T16:03:47.000Z"
+
+    mocker.patch.object(extractor.state_store, "get_state", return_value=(last_update_time,))
+    mocker.patch.object(extractor.client.time_series.data, "insert_dataframe", return_value=None)
+    mocker.patch("cdf_fabric_replicator.extractor.CdfFabricExtractor.set_state", return_value=None)
+
+    extractor.write_time_series_to_cdf(df)
+
+    expected_update_list = df[df["timestamp"] > last_update_time]
+
+    extractor.state_store.get_state.assert_any_call("/table/path-id1-state")
+    assert extractor.client.time_series.data.insert_dataframe.call_count == 1
+    assert len(extractor.client.time_series.data.insert_dataframe.call_args_list[0]) == len(expected_update_list)
+    extractor.set_state.assert_any_call("/table/path-id1-state", df[df["externalId"] == "id1"]["timestamp"].max())
+
+
+def test_write_time_series_to_cdf_no_new_data_points(extractor, mocker):
+    data = {
+        "externalId": ["id1", "id1", "id1", "id1"],
+        "timestamp": [
+            "2022-02-07T16:01:27.001Z",
+            "2022-02-07T16:03:47.000Z",
+            "2022-02-07T16:03:52.000Z",
+            "2022-02-07T16:04:12.000Z",
+        ],
+        "value": [1.0, 2.0, 3.0, 4.0],
+    }
+    df = pd.DataFrame(data)
+
+    last_update_time = "2022-02-07T16:04:12.000Z"
```

Comment:
this seems to be the same value as the last value in the timestamp array above. Should they both be initialized from the same const to tell the reader that those need to be the same?

Reply:
It doesn't have to be the same timestamp. What we tried to test here is given the timestamp of last data that we've processed (from the state store), if some data has an earlier timestamp and some has a later timestamp, we will only write the latter to CDF.

Reply:
thank you that's a good point. done